### ISSUE TITLE
Properly catch OperationalError 

### ISSUE THREAD

**Issue by philpep** [psycopg/psycopg#242]
Hi,

I'm trying to catch `OperationalError('connection failed: fe_sendauth: no password supplied')`

Exception message is unreliable for this since it might be localized. And I see no other field on exception to filter on:

```
>>> exc.sqlstate
None
```

And all fields on `exc.diag` are None, looks like a bug no ?

This is on `psycopg 3.0.9`

Thanks!

**Comment by dvarrazzo** [comment]
No, it's not a bug. sqlstate, as well as other error fields, are not available for any error. The specific error you mention is generated by the client library, not by the server, and I doubt that the client returns any SQLSTATE (if it does, it is more an exception than a rule).

Looking at the libpq code, it doesn't seem that the message in question undergoes any l10n

```
piro@baloo:~/dev/fs/postgres$ ag -C2 PQnoPasswordSupplied src/interfaces/libpq/
src/interfaces/libpq/libpq-fe.h
557-/* Error when no password was given. */
558-/* Note: depending on this is deprecated; use PQconnectionNeedsPassword(). */
559:#define PQnoPasswordSupplied	"fe_sendauth: no password supplied\n"
560-
561-/* Create and manipulate PGresults */

src/interfaces/libpq/fe-auth.c
559-	{
560-		appendPQExpBufferStr(&conn->errorMessage,
561:							 PQnoPasswordSupplied);
562-		goto error;
563-	}
--
1031-				{
1032-					appendPQExpBufferStr(&conn->errorMessage,
1033:										 PQnoPasswordSupplied);
1034-					return STATUS_ERROR;
1035-				}
```

However relying on the message is certainly brittle.

The `PGconn` object offers functions to detect if a password was required but not provided. This interface is stable, although not documented yet (the `needs_password` attribute is a wrapper for the `PQconnectionNeedsPassword()` mentioned in the libpq comment above):

```python
>>> cnn = psycopg.pq.PGconn.connect(b"host=localhost user=foo")
>>> cnn
<psycopg_c.pq.PGconn [BAD] at 0x7f49d78db310>

>>> psycopg.pq.misc.error_message(cnn)
'fe_sendauth: no password supplied'

cnn.needs_password
True

cnn.password
b''
```

However, when `Connection.connect()` fails, the underlying `PGconn` object is discarded. We might attach the `PGconn` to the exception object instead, to make it available in this case, but this is a new feature, not a bugfix.

Is this an interface you would see yourself using? We might add it in psycopg 3.1.

**Comment by dvarrazzo** [comment]
I have put together a proof of concept in the [pgconn-in-exception](https://github.com/psycopg/psycopg/tree/pgconn-in-exception) branch. adding a `pgconn` attribute on the exception, which gets populated only when there is an exception on connection.

Would you like to take a look if, using this feature, you are able to tell apart the error conditions you need?

You can take a look at the [`PQconnectionNeedsPassword()` and `PQconnectionUsedPassword()` docs](https://www.postgresql.org/docs/current/libpq-status.html#LIBPQ-PQCONNECTIONNEEDSPASSWORD), wrapped as `PGconn.needs_password` and `PGconn.used_password`.

**Comment by dvarrazzo** [comment]
Hi @philpep,

we have added information on the exception object which should allow to investigate better on failed connection causes, without the need of parsing the error message.

It should be released in Psycopg 3.1. We will try to figure out, in the next days, when to release it. There are some works still in progress, but we have gathered enough new features that we can consider to release less and earlier, and postpone something to 3.2.

Thank you very much @dlax for reviewing the approach taken to solve this type of problems :slightly_smiling_face:
